### PR TITLE
CONSOLE-2976: add update mode to Update cluster

### PR DESCRIPTION
**Pull request by rhamilto**
This functionality is desired to be back ported all the way to 4.8.  As a result, it uses the following deprecated functions for ease of back porting:

* useAccessReview
* k8sPatch

It adds:
* the `Update options` radio options to the `Update cluster` modal
* the ability to pause/unpause worker or custom pools within each pool's progress bar after the update has started
* the alert to the top of the `Cluster Settings` page if one or more worker or custom pools is paused

TODOs:

- [x] get doc url
- [ ] update pause message once https://issues.redhat.com/browse/MCO-77 lands (to be added with [CONSOLE-3126](https://issues.redhat.com/browse/CONSOLE-3126))

https://user-images.githubusercontent.com/895728/159528245-bd6ae823-3abc-4339-b014-419046bef75d.mov

https://user-images.githubusercontent.com/895728/159528253-2de600f4-d643-4766-a60b-47b87980711f.mov

Screen Shot 2022-02-14 at 2 00 38 PM

https://user-images.githubusercontent.com/895728/153949126-cfa7f52c-c40d-4d4a-a1b2-9056ba2916d8.mov

Screen Shot 2022-02-21 at 4 06 02 PM

### CODE REVIEW

**Comment by spadgett** [comment]
@sdodson FYI

**Comment by sdodson** [comment]
The docs that address EUS-to-EUS upgrades are at https://docs.openshift.com/container-platform/4.9/updating/preparing-eus-eus-upgrade.html

We will want to make sure that we update those with the new web UI work here and we'll need to make sure that they guide the admin toward
1. Partial 4.10 to 4.11
2. Partial 4.11 to 4.12
3. UnPause

If we let them do the following they'd immediately unpause pools and workers, infra, etc will end up updating from 4.10 to 4.11 while the rest of the cluster is upgrading from 4.11 to 4.12. Nothing should break, but it will defeat the intent of skipping the worker updates between 4.10 and 4.11.
1. Partial 4.10 to 4.11
2. Full 4.11 to 4.12

I'm not sure how we could improve the messaging in console to address that, perhaps just put emphasis on the fact that pools will be unpaused immediately when you click a full upgrade?

**Comment by julienlim** [comment]
Overall the flow looks fine, though I'd suggest the text be reviewed as there's room for improvement there and minor typos. 

Here are some examples (there's others but I focused only the first screen):

"Both of master and worker nodes are updated at once. This might take longer, so make sure you allocate enough time for maintenance schedule." --> "Both master and worker nodes are updated at once. This may take awhile, so allocate enough time for maintenance."

"You can pause any non-master node updates to accommodate your maintenance schedule." --> "Pause any worker or infra nodes to accommodate your maintenance schedule."

"Nodes" --> "nodes"

**Comment by rhamilto** [comment]
@sdodson, @julienlim Thanks for the reviews!  I will make sure the designers see your feedback and it is addressed.

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (rhamilto) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:173]: This annotation is new as of 4.10 and will need to be back ported

**Comment by sdodson** [comment]
@rhamilto Is it easy to include how long pools have been paused, either in initial implementation or via a future story?

**Comment by rhamilto** [comment]
QE Approver:
/assign @yapei

Docs Approver:
/assign @opayne1

PX Approver:
/assign @RickJWagner

**Comment by rhamilto** [comment]
@jhadvig, @spadgett PTAL when you have a chance (note this is not pressing since we're in violation of SLO and are focused on bugs).

**Comment by yapei** [comment]
@rhamilto  @yanpzhan is testing the PR and will report back ~

**Comment by rhamilto** [comment]
Thanks, @yanpzhan.  The test case is approved.

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (andrewballantyne) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:382]: Is this an assumption we should be making? Can't RBAC checks deny access partially to some resources? Are we assuming this will always be a super admin so if they have one access they'll have all?

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (andrewballantyne) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:661]: Do we have a client-side ticket to handle this? MCO-77 is an Epic so do we want to create a story on our side / on their side to adjust this to have the right value?

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (andrewballantyne) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:859]: These (this one and the Worker one below) are not translated. They appear to be used directly in both data and text rendered to the screen in `NodesUpdatesGroup`. We should consider splitting them up and translating the visual one and keeping the data one separate.

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (andrewballantyne) [inline frontend/public/components/modals/cluster-update-modal.tsx:81]: What if `id` is not available... you go from a value to undefined. Is this okay?

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (andrewballantyne) [inline frontend/public/components/modals/cluster-update-modal.tsx:69]: What are you pulling? I'm pretty sure `_.pull` does not do anything by itself. Also `mcp.metadata.name` is a string. Whereas [the docs](https://lodash.com/docs/4.17.15#pull) say it pulls from arrays.

Did you mean to do:
```
_.pull(initialMCPPausedValues, mcp.metadata.name);
```

Comment (andrewballantyne) [inline frontend/public/components/modals/cluster-update-modal.tsx:69]: You shouldn't really ever do a `.push` on an array in the middle of a `.map`. What you want is `.forEach` I think in this case... Personally I favour a `.reduce` as you're building an array that can dynamically grow and shrink as you iterate over the data. 

If you're just looking at only keeping values that are not master & are paused... I think you just want a `.filter`, no?

File: frontend/public/module/k8s/cluster-settings.ts
Comment (andrewballantyne) [inline frontend/public/module/k8s/cluster-settings.ts:292]: Two notes here:

1. You shouldn't really mutate properties that are passed to you -- you should always create your own and return the value; this is because it's harder to track mutations when reading code than it is to read joining of values together.
2. `.map` always expects a value... not returning means you probably want a `.forEach`; JavaScript allows no returns, but our TypeScript/Linter (outside of public folder) will have a harder time with it (or it should for code quality)

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (andrewballantyne) [inline frontend/public/components/modals/cluster-update-modal.tsx:179]: NodeTypeNames are not translated... any reason we are not making the read values part of the translations? Are these values expected to change? I would think `Master` and `Worker` are idiomatic names and thus not subject to changes.

File: frontend/__tests__/components/cluster-settings.spec.tsx
Comment (andrewballantyne) [inline frontend/__tests__/components/cluster-settings.spec.tsx:267]: This will always be true...

See this issue for an official response about Enzyme: https://github.com/enzymejs/enzyme/issues/1864#issuecomment-430113911

**Review by andrewballantyne — CHANGES_REQUESTED** [review]

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (rhamilto) [reply frontend/public/components/cluster-settings/cluster-settings.tsx:382]: Yes to the assumption the user is a super admin.  We don't know about the others because they are optional.

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (rhamilto) [reply frontend/public/components/cluster-settings/cluster-settings.tsx:661]: MCO-77 was the issue I was referred to by @jkyros when we discussed the need for the annotation.  Not sure how he is tracking the work on his side.  See https://issues.redhat.com/browse/MCO-77?focusedCommentId=19968846&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-19968846

Comment (andrewballantyne) [reply frontend/public/components/cluster-settings/cluster-settings.tsx:661]: Right... but unless you're holding your PR up for this backend change, we need a ticket to manage this UI change. So we should create something so it does not get lost. If you're unsure where to create it, I suggest the CONSOLE project. Maybe on the parent epic? https://issues.redhat.com/browse/CONSOLE-2927

Doesn't hurt to create a placeholder ticket so we don't lose this in the cracks. If you are waiting for it before we merge, perhaps put a `/hold` on your PR with that as a note.

Comment (rhamilto) [reply frontend/public/components/cluster-settings/cluster-settings.tsx:661]: Good point.  Added [CONSOLE-3126](https://issues.redhat.com/browse/CONSOLE-3126).

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (rhamilto) [reply frontend/public/components/modals/cluster-update-modal.tsx:69]: Thanks.  Will change that `map` to a `forEach` as in this case I'm creating a new array of strings (going to add a type on line 65 also) in order to set which MCPs are paused when the modal is launched (if any).  In theory the `pull` will never occur because the array is initialized to empty each to the effect runs, but figured it was a good measure.  Is there another approach I should consider?

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (rhamilto) [reply frontend/public/components/modals/cluster-update-modal.tsx:81]: In theory it shouldn't.  Are you raising the concern because of the (probably unnecessary) optional chaining on 80?  What is the correct approach?

File: frontend/public/module/k8s/cluster-settings.ts
Comment (rhamilto) [reply frontend/public/module/k8s/cluster-settings.ts:292]: > You shouldn't really mutate properties that are passed to you

In this case, `promises`?

> you probably want a .forEach

Indeed.  Will change.

**Comment by rhamilto** [comment]
@andrewballantyne, I made an attempt at addressing your feedback with 6cd2af4d5f048c8d74377b581eb62ca81e26981d.  Will squash once everything is resolved.

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (andrewballantyne) [reply frontend/public/components/modals/cluster-update-modal.tsx]: If you're looking to take a list of resources, turn it into a list of their `.metadata.name`s and omit anything that doesn't fit a given state (ie `!isMCPMaster(mcp) && isMCPPaused(mcp)`) there is a simple approach: `[].filter().map()`:

```
const initialMCPPausedValues = machineConfigPools
  .filter((mcp) => !isMCPMaster(mcp) && isMCPPaused(mcp))
  .map((mcp) => mcp.metadata.name)
```
Which is easy to read as "taking machineConfigPool resources, filter ones we don't care about, just get the name of the resource". Now you have a simple list of strings ready to go.

Comment (andrewballantyne) [reply frontend/public/components/modals/cluster-update-modal.tsx]: The reason I didn't suggest this earlier was I was not sure if your `pull` was intending to try and de-duplicate values that were previously added in a previous loop of `.map`.

File: frontend/public/module/k8s/cluster-settings.ts
Comment (andrewballantyne) [reply frontend/public/module/k8s/cluster-settings.ts]: That is correct, `promises` is mutated when you call `.push`. So the caller passes you an array and you mutate it. It's anti-pattern for functional programming to mutate a calling variable. In this case, you're doing both sides of the fence so you "know" that's the whole point of passing the array. But as code mutates over time, it gets less and less clear and the code is likely not to be easily reused.

`.forEach` is for side-effects... in this case your mutation of promises. But the most important is to avoid `.map` that never returns... because JS will return undefined... and then your method now returns an array of undefined equal to the length of `machineConfigPools`. Which is definitely not what you want to do.

File: frontend/public/module/k8s/cluster-settings.ts
Comment (andrewballantyne) [inline frontend/public/module/k8s/cluster-settings.ts:293]: Ah, haha... so sorry. This is my fault for trying to explain everything and not being clear about what I was wanting.

`.map` is fine, but you need to return out of the predicate. Not mutating is better than mutating. So your current state really should be a middle ground between what you had and what you have now. Again, this is my bad for explaining poorly.

```suggestion
  return machineConfigPools.map((mcp) => {
    const patch = [{ op: 'add', path: '/spec/paused', value: paused }];
    return k8sPatch(MachineConfigPoolModel, mcp, patch);
  });
```

File: frontend/__tests__/components/cluster-settings.spec.tsx
Comment (andrewballantyne) [inline frontend/__tests__/components/cluster-settings.spec.tsx:245]: Small nit. `React.ComponentProps<typeof Component>` is how you get props, instead of `any`.

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (rhamilto) [reply frontend/public/components/modals/cluster-update-modal.tsx]: Updated to your suggestion and limited to effect to run once.  It occurred to me we don't want to change the checkboxes once they are rendered because that could inadvertently override user input.

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (andrewballantyne) [reply frontend/public/components/modals/cluster-update-modal.tsx:79]: `Radio` is a PF component. `event` will not be `undefined` or `null` (or PF has a bug, not us). `currentTarget` is always going to be there because it's an event. `id` is set by your component, which I'll admit is not the best for React components (`id`s are intended to be unique in the whole HTML document and thus don't play nicely with React components that can render multiple times). So they are entirely not optionally chained because it'll all be there.

However, my recommendation is if you're unsure of where this data is coming from or how you want to deliver such a piece of information in a guaranteed fashion... use closures so you have full control.

Eg make your component:
```
<Radio onChange={(checked) => handleRadioOnChange(checked, SOME_CONSTANT)} />
```
That way you have just wrapped your callback in the value it needs without trying to pass a value through the DOM to your own callback.

**Review by andrewballantyne — COMMENTED** [review]
Missed a comment from Friday.

Also, you didn't address it directly, but `Worker` and `Master` are not translated, right? Perhaps we put a comment next to the enum to make sure no one makes that mistake in the future.

**Comment by rhamilto** [comment]
/hold

found some bugs

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (rhamilto) [reply frontend/public/components/modals/cluster-update-modal.tsx:166]: Good catch.  That was an oversight.

File: frontend/public/components/modals/cluster-update-modal.tsx
Comment (rhamilto) [reply frontend/public/components/modals/cluster-update-modal.tsx:180]: `id` is a required prop.  See https://www.patternfly.org/v4/components/radio#radio

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (andrewballantyne) [inline frontend/public/components/cluster-settings/cluster-settings.tsx:385]: Do we use the constant here?

**Review by andrewballantyne — COMMENTED** [review]
Any reason why we go directly to full upgrade and prompt this message in the middle of a partial upgrade?

image

File: frontend/public/components/cluster-settings/cluster-settings.tsx
Comment (rhamilto) [reply frontend/public/components/cluster-settings/cluster-settings.tsx]: Good catch.  Updated.

**Comment by rhamilto** [comment]
> Any reason why we go directly to full upgrade and prompt this message in the middle of a partial upgrade?
> 
> image

Full is the default, and there is no requirement that the previous update be complete before starting a new one.

**Comment by rhamilto** [comment]
@opayne1, the back port of this feature to 4.10.z will likely occur before the release of 4.11.  Are there any doc needs for 4.10.z that should come before the back port?

**Pull request merged**
